Sources/Networking/Modifiers/Interceptors/Authorization/AuthorizationTokenInterceptor.swift

```diff
@@ -0,0 +1,104 @@
+//
+//  AuthorizationTokenInterceptor.swift
+//  
+//
+//  Created by Dominika Gajdová on 08.12.2022.
+//
+
+import Foundation
+
+// MARK: - Defines authentication handling in requests
+public final class AuthorizationTokenInterceptor: RequestInterceptor {
+    private let refreshingState = RefreshingState()
+    private var authorizationManager: AuthorizationManaging
+    
+    public init(authorizationManager: AuthorizationManaging) {
+        self.authorizationManager = authorizationManager
+    }
+    
+    public func adapt(_ request: URLRequest, for endpointRequest: EndpointRequest) async throws -> URLRequest {
+        guard endpointRequest.endpoint.isAuthenticationRequired else {
+            return request
+        }
+        
+        do {
+            return try await authorizationManager.authorizeRequest(request)
+        } catch {
+            /// If authorization fails due to expiredAccessToken we should perform refresh and then retry the request adaptation again.
+            guard case AuthorizationError.expiredAccessToken = error else {
+                throw error
+            }
+            
+            try await performRefresh()
+            
+            return try await adapt(request, for: endpointRequest)
+        }
+    }
+    
+    public func process(_ response: Response, with urlRequest: URLRequest, for endpointRequest: EndpointRequest) async throws -> Response {
+        guard let httpResponse = response.response as? HTTPURLResponse else {
+            throw NetworkError.noStatusCode(response: response)
+        }
+        
+        /// Request was unauthorized but required valid authorization.
+        guard
+            httpResponse.statusCode == 401,
+            endpointRequest.endpoint.isAuthenticationRequired
+        else {
+            return response
+        }
+        
```

Comment:
True. @gajddo00 Can you tell us what is this supposed to do? I actually think that this class should be just an Adapter instead of Interceptor 🧐

Reply:
yes we should refresh here I believe

Reply:
I don't think it makes sense to refresh here, since we can't retry the call afterwards from a processor function. If we just let it fail and it is going to be retried by the APIManager, the token will get refreshed in the adapter function. That's why I believe this class doesn't need to be an Interceptor but rather an Adapter.

Reply:
Oh right, makes sense. Implemented ✅